DDL Checker Tool

zier-one opened the pull request:
### What problem does this PR solve? 

add a tool can check if DDL SQL can be successfully executed by TiDB.

### What is changed and how it works?

run a part of tidb,and execute sql to check if it can run.also,this tool can synchronize the dependent table structure from MYSQL automatically. 

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - [x] Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

holys commented:
LGTM

andelf commented on pkg/tableroute/trieroute.go:
```diff
@@ -0,0 +1,252 @@
+// Copyright 2016 PingCAP, Inc.
```
2017

andelf commented on pkg/tableroute/trieroute.go:
```diff
@@ -0,0 +1,252 @@
+	AllRules() map[string]string
+}
+
+type trieRouter struct {
```
add wikipedia link for this type of data structure

andelf commented on pkg/tableroute/trieroute.go:
```diff
@@ -0,0 +1,252 @@
+}
+
+// NewTrieRouter returns a trie Router
+func NewTrieRouter() Router {
```
New()

GregoryIan replied:
as the interface defined above, maybe there are other implements

andelf commented on pkg/tableroute/trieroute.go:
```diff
@@ -0,0 +1,252 @@
+		return errors.Errorf("pattern %s and target %s can't be empty", pattern, target)
+	}
+	if pattern[0] == asterisk {
+		return errors.Errorf("invalid pattern %s", pattern)
```
unsupported pattern

andelf commented on pkg/tableroute/trieroute.go:
```diff
@@ -0,0 +1,252 @@
+	t.Lock()
+
+	n := t.root
+	hadAsterisk := false
```
has

andelf commented on pkg/tableroute/trieroute.go:
```diff
@@ -0,0 +1,252 @@
+// 2. the question mark ? matches exactly one character
+const (
+	// asterisk [ * ]
+	asterisk = '*'
```
upper case

holys replied:
@andelf  Why upper case?

andelf replied:
const

hicqu replied:
private variables shoud be lower case?

GregoryIan commented:
update: the route rule in syncer would be like
```
schema = "schema_*"
table = "abc_*"
pattern-schema = "schema"
pattern-table = "abc"
```
this rule would route [schema_1, abc_1] to [schema, abc] .
the detail would be in next pr(synced supports table route)
@iamxy @holys PTAL

holys commented on pkg/tableroute/trieroute_test.go:
```diff
@@ -0,0 +1,71 @@
+func (t *testRouteSuite) TestRoute(c *C) {
+	r := NewTrieRouter()
+	t.testInsert(c, r)
+	//t.testMatch(c, r)
```
Why comment here?

hicqu commented:
LGTM.

holys commented:
LGTM

Commit:
ddl checker tool

	修改：      Makefile
	新文件：    ddl_checker/main.go
	修改：      go.mod
	修改：      go.sum
	新文件：    pkg/ddl-checker/ddl_checker.go
	新文件：    pkg/ddl-checker/ddl_checker_test.go

Commit:
add README

kennytm commented on ddl_checker/README.md:
```diff
@@ -0,0 +1,58 @@
+
+cd ddl_checker
+go build
+./ddl_checker --host [host] --port [port] --user [user] --password [password] --schema [schrma]
```
```suggestion
./ddl_checker --host [host] --port [port] --user [user] --password [password] --schema [schema]
```

kennytm commented on ddl_checker/README.md:
```diff
@@ -0,0 +1,58 @@
+Auto mode: The program will automatically synchronize the dependent table structure from MYSQL and delete the conflict table.   
+Query mode: The program will ask you before synchronizing the dependent table structure from MYSQL.  
+Manual mode: This program does not perform anything other than executing the input SQL.  
+SETMOD Usage: SETMOD <MODCODE>; MODCODE = [0(Auto), 1(Query), 2(Manual)] 
```
Could the tool be made to recognize the strings "auto", "query" and "manual" instead of 0, 1, 2?

BTW usually we use "prompt" or "ask". The term "query" is confusing esp. when we're talking about SQL.

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,216 @@
+	initialise()
+	printWelcome()
+	mainLoop()
+	//handler("ALTER TABLE table_name MODIFY column_1 int(11) NOT NULL;")
```
Delete this comment?

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,216 @@
+}
+func handler(input string) bool {
+	// cmd exit
+	lowerTrimInput := strings.ToLower(strings.TrimSpace(input[:len(input)-1]))
```
Use `strings.TrimRight(input, "; \t\n\r")` to removing the trailing semicolon. (or `TrimRightFunc` if you do care about Unicode)

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,216 @@
+		nonNeededTable := ddl_checker.GetTableNeededNonExist(stmt)
+		// query mod
+		if modCode == 1 && !queryAutoSync(neededTable, nonNeededTable) {
+			goto EXECUTE
```
wut

there's no need to use `goto` here. you could simply invert the condition

```go
if modCode != 1 || queryAutoSync(neededTable, nonNeededTable) {
    syncTablesFromMysql(neededTable)
    dropTables(nonNeededTable)
}
```

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,216 @@
+	reader = bufio.NewReader(os.Stdin)
+	executableChecker, err = ddl_checker.NewExecutableChecker()
+	if err != nil {
+		fmt.Printf("[DDLChecker]Init failed,Can't create ExecutableChecker: %s\n", err.Error())
```
```suggestion
		fmt.Printf("[DDLChecker] Init failed, can't create ExecutableChecker: %s\n", err.Error())
```

(similar for other messages: there should be a space after the comma `,` and you don't need to capitalize the word after the comma)

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,216 @@
+			fmt.Println("[DDLChecker] Table", tableName, "is exist,Skip")
+			continue
+		}
+		rows, err := db.Query("show create table " + tableName)
```
1. you could use `QueryRow` here
2. you need to escape the table name, at the minimum:

    ```go
    "SHOW CREATE TABLE `" + tableName + "`"
    ```

    there might  be an existing function here to add the quotation marks.

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,216 @@
+		case "n":
+			return false
+		case "":
+			goto QUERY
```
This `goto` isn't really needed besides some cosmetic details.

kennytm commented on pkg/ddl-checker/ddl_checker.go:
```diff
@@ -0,0 +1,123 @@
+
+func (ec *ExecutableChecker) IsTableExist(tableName string) bool {
+	_, err := ec.session.Execute(ec.context,
+		fmt.Sprintf("select 0 from %s limit 1", tableName))
```
Quote the table name

kennytm commented on pkg/ddl-checker/ddl_checker_test.go:
```diff
@@ -0,0 +1,93 @@
+		tableNeededExist := GetTableNeededExist(stmt)
+		tableNeededNonExist := GetTableNeededNonExist(stmt)
+		c.Assert(data.parseSucceeded, IsTrue)
+		c.Assert(reflect.DeepEqual(data.tableNeededExist, tableNeededExist), IsTrue)
```
```suggestion
		c.Assert(data.tableNeededExist, DeepEquals, tableNeededExist)
```

Similar to the rest.

Commit:
take kenny's advice,part 1

Commit:
take kenny's advice,part 2

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,212 @@
+			case "":
+				continue
+			default:
+				break
```
the `break` only breaks the `switch`, maybe you want to break to a label?

https://stackoverflow.com/questions/11104085/in-go-does-a-break-statement-break-from-a-switch-select

zier-one replied:
oh,it's my mistake

Commit:
take kenny's advice,part 3

kennytm commented on ddl_checker/README.md:
```diff
@@ -0,0 +1,58 @@
+[DDLChecker] Syncing Table table_name
+[DDLChecker] SQL execution succeeded
+
+[Auto] > SETMOD 1;
```
```suggestion
[Auto] > SETMOD PROMPT;
```

kennytm commented on ddl_checker/README.md:
```diff
@@ -0,0 +1,58 @@
+[DDLChecker] Table table_name is exist,Skip
+[DDLChecker] SQL execution succeeded
+
+[Query] > setmod 2;
```
```suggestion
[Query] > setmod manual;
```

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,214 @@
+		fmt.Println("[DDLChecker] Syncing Table", tableName)
+		isExist := executableChecker.IsTableExist(tableName)
+		if isExist {
+			fmt.Println("[DDLChecker] Table", tableName, "is exist,Skip")
```
```suggestion
			fmt.Println("[DDLChecker] Table", tableName, "exists, skipping")
```

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,214 @@
+func dropTables(tableNames []string) {
+	for _, tableName := range tableNames {
+		fmt.Println("[DDLChecker] Dropping table", tableName)
+		err := executableChecker.Execute(fmt.Sprintf("drop table if exists %s", tableName))
```
```suggestion
		err := executableChecker.Execute(fmt.Sprintf("drop table if exists `%s`", tableName))
```

kennytm commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,214 @@
+func modeName() string {
+	switch modCode {
+	case 0:
+		return " Auto "
```
why the spaces around "Auto" 🤔 

```suggestion
		return "Auto"
```

kennytm reviewed (commented):
LGTM.

Commit:
take kenny's advice,part 4

IANTHEREAL commented:
remove useless description
```Code changes

Has exported function/method change
Has exported variable/fields change
Has interface methods change
Has persistent data change
Side effects

Possible performance regression
Increased code complexity
Breaking backward compatibility
Related changes

Need to cherry-pick to the release branch
Need to update the documentation
Need to be included in the release note```

IANTHEREAL commented on ddl_checker/README.md:
````diff
@@ -0,0 +1,58 @@
+## How to use
+
+```
+Usage of checker:
````
s/checker/ddl_checker

IANTHEREAL commented on ddl_checker/README.md:
```diff
@@ -0,0 +1,58 @@
+[DDLChecker] Table table_name exist, skipping
+[DDLChecker] SQL execution succeeded
+
+[Query] > setmod manual;
```
could it support offline mode?   users don't need to input `host`,`port`.. and input create tabel and ddl statement by themself

zier-one replied:
good idea

IANTHEREAL commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,214 @@
+
+var (
+	modCode                                          = 0
+	executableChecker *ddl_checker.ExecutableChecker = nil
```
no need initial value

IANTHEREAL commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,214 @@
+func mainLoop() {
+	var input string
+	var err error
+	for cont := true; cont; cont = handler(input) {
```
do you want to use `github.com/spf13/cobra`?

IANTHEREAL commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,214 @@
+	}
+}
+func handler(input string) bool {
+	// cmd exit
```
put it before L100

IANTHEREAL commented on pkg/ddl-checker/ddl_checker.go:
```diff
@@ -0,0 +1,122 @@
+
+func (ec *ExecutableChecker) Parse(sql string) (stmt ast.StmtNode, err error) {
+	charset, collation := ec.session.GetSessionVars().GetCharsetInfo()
+	stmt, err = ec.parser.ParseOneStmt(sql, charset, collation)
```
you need a `NewValueExpr` at https://github.com/pingcap/parser/blob/master/ast/expressions.go#L63. otherwise ddl_check would panic if it meets some statement contain` new value expression`

IANTHEREAL commented on pkg/ddl-checker/ddl_checker.go:
```diff
@@ -0,0 +1,122 @@
+	case *ast.RenameTableStmt:
+		return []string{x.OldTable.Name.String()}
+	default:
+		return []string{}
```
`return nil`

IANTHEREAL commented on pkg/ddl-checker/ddl_checker.go:
```diff
@@ -0,0 +1,122 @@
+	case *ast.RenameTableStmt:
+		return []string{x.NewTable.Name.String()}
+	default:
+		return []string{}
```
ditto

IANTHEREAL commented on pkg/ddl-checker/ddl_checker.go:
```diff
@@ -0,0 +1,122 @@
+	}
+}
+
+func IsDDL(stmt ast.StmtNode) bool {
```
```
switch stmt.(type) {
	case ast.DDLNode:
```

IANTHEREAL commented on pkg/ddl-checker/ddl_checker.go:
```diff
@@ -0,0 +1,122 @@
+	return
+}
+
+func GetTableNeededExist(stmt ast.StmtNode) []string {
```
s/GetTableNeededExist/GetTablesNeededExist

IANTHEREAL commented on pkg/ddl-checker/ddl_checker.go:
```diff
@@ -0,0 +1,122 @@
+	}
+}
+
+func GetTableNeededNonExist(stmt ast.StmtNode) []string {
```
s/GetTableNeededNonExist/GetTablesNeededNonExist

IANTHEREAL commented on pkg/ddl-checker/ddl_checker.go:
```diff
@@ -0,0 +1,122 @@
+	parser  *parser.Parser
+}
+
+func NewExecutableChecker() (*ExecutableChecker, error) {
```
how to use it in other tools like dm-ctl?   need to reproduce logic in `ddl_checker/main.go`？

zier-one replied:
yes,If we need to reuse the functionality that sync tables from mysql, I should refactor the code

IANTHEREAL commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,214 @@
+func syncTablesFromMysql(tableNames []string) {
+	for _, tableName := range tableNames {
+		fmt.Println("[DDLChecker] Syncing Table", tableName)
+		isExist := executableChecker.IsTableExist(tableName)
```
maybe not user want？

zier-one replied:
should i ask the user every table?

IANTHEREAL replied:
maybe local table is not the table user want, just drop it and create a new one

zier-one replied:
this tool will skip syncing the table what already exist.when user drop the local table,the tool  will just drop it

IANTHEREAL commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,214 @@
+		neededTable := ddl_checker.GetTableNeededExist(stmt)
+		nonNeededTable := ddl_checker.GetTableNeededNonExist(stmt)
+		if modCode == 0 || (modCode == 1 && queryAutoSync(neededTable, nonNeededTable)) {
+			syncTablesFromMysql(neededTable)
```
if some errors happen in these two function `syncTablesFromMysql` and `dropTables`, would it continue to execute?

zier-one replied:
yes,it will print error info,and continue to try executing the input sql.

IANTHEREAL replied:
would it exit?

zier-one replied:
it wonldn't

Commit:
take ian's advice,part 1

Commit:
take ian's advice,part 2

Commit:
take ian's advice,part 3

Commit:
fix skip when stmt isn't a DDLNode

Commit:
resolve conflicts

IANTHEREAL commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,222 @@
+func mainLoop() {
+	var input string
+	var err error
+	for cont := true; cont; cont = handler(input) {
```
what's meaning of  `cont`?

zier-one replied:
continue; control the main loop continue or exit

zier-one replied:
maybe the var name is not good enough.

IANTHEREAL replied:
it's sure

Commit:
fix cli logic

Commit:
fix cli logic

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
fix cli logic

Commit:
refactor ddl_checker

IANTHEREAL commented on ddl_checker/README.md:
```diff
@@ -0,0 +1,59 @@
+[Query] > ALTER TABLE table_name MODIFY column_1 int(11) NOT NULL;
+[DDLChecker] Do you want to synchronize table [table_name] from MySQL and drop table [] in ExecutableChecker?(Y/N)y
+[DDLChecker] Syncing Table table_name
+[DDLChecker] Table table_name exist, skipping
```
remove it?

IANTHEREAL commented on pkg/ddl-checker/ddl_syncer.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func (ds *DDLSyncer) SyncTable(schemaName string, tableName string) error {
+	err := ds.ec.DropTable(tableName)
```
first get then drop? but it doesn't matter, only a bit better

IANTHEREAL commented on pkg/ddl-checker/ddl_syncer.go:
```diff
@@ -0,0 +1,49 @@
+package ddl_checker
```
add license

IANTHEREAL commented on pkg/ddl-checker/executable_checker.go:
```diff
@@ -0,0 +1,129 @@
+
+type ExecutableChecker struct {
+	session  session.Session
+	context  context.Context
```
try to avoid putting the context in the structure,  the best way to use context is to initialize one that accompanies the process at the beginning of the process

IANTHEREAL commented on pkg/ddl-checker/executable_checker.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func (ec *ExecutableChecker) DropTable(tableName string) error {
+	err := ec.Execute(fmt.Sprintf("drop table if exists `%s`", tableName))
```
ExecuteContext may be better

IANTHEREAL commented on pkg/ddl-checker/executable_checker.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func (ec *ExecutableChecker) Close() error {
+	if ec.isClosed {
```
atomic compare and swap？

IANTHEREAL commented on pkg/ddl-checker/ddl_syncer.go:
```diff
@@ -0,0 +1,49 @@
+	return &DDLSyncer{db, executableChecker}, nil
+}
+
+func (ds *DDLSyncer) SyncTable(schemaName string, tableName string) error {
```
add  comments for exported function, doesn't your go lint output errors?

zier-one replied:
goline is commented out in makefile

IANTHEREAL replied:
uncomment it

IANTHEREAL commented on pkg/ddl-checker/executable_checker.go:
```diff
@@ -0,0 +1,129 @@
+	return err == nil
+}
+
+func (ec *ExecutableChecker) DropTable(tableName string) error {
```
add a `CreateTable` function?

IANTHEREAL commented:
Rest LGTM

Commit:
take ian's advice,part 4

IANTHEREAL commented on pkg/ddl-checker/executable_checker.go:
```diff
@@ -49,47 +51,64 @@ func NewExecutableChecker() (*ExecutableChecker, error) {
+// Close closes the ExecutableChecker
 func (ec *ExecutableChecker) Close() error {
-	if ec.isClosed {
+	if ec.isClosed.Load() || !ec.isClosed.CAS(false, true) {
```
`if !ec.isClosed.CAS(false, true)` is enough

Commit:
take ian's advice,part 5

IANTHEREAL commented:
@amyangfei @csuzhangxc PTAL

IANTHEREAL commented:
LGTM

amyangfei commented on pkg/ddl-checker/executable_checker.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+// GetTablesNeededNonExist reports the table name that conflicts with ast.StmtNode
+// the specified ast.StmtNode must a DDLNode
```
s/must/must be

amyangfei commented on pkg/ddl-checker/executable_checker.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+// GetTablesNeededExist reports the table name needed to execute ast.StmtNode
+// the specified ast.StmtNode must a DDLNode
```
ditto

amyangfei commented on ddl_checker/main.go:
```diff
@@ -0,0 +1,209 @@
+			}
+		}
+	}
+	err := executableChecker.Execute(tidbContext, input)
```
If I input a valid non-DDL SQL, the checker will tell me `SQL execution succeeded`. The tool is more like a `SQL checker` with a TiDB context. Maybe we can put some DDL check here and tell user whether the input SQL is a DDL or not, but this requires SQL parsing and some more work, which is not as simple as current code. But I think it is OK now.

zier-one replied:
i can warning users when they  input a valid non-DDL SQL and continue execute the sql.

amyangfei reviewed (commented):
Rest LGTM

Commit:
take amyangfei's advice,part 1

amyangfei reviewed (commented):
LGTM
